src/openzeppelin/account.cairo

```diff
@@ -0,0 +1,233 @@
+use starknet::ContractAddress;
+use openzeppelin::utils::check_gas;
+
+const ERC165_ACCOUNT_ID: u32 = 0xa66bd575_u32;
+const ERC1271_VALIDATED: u32 = 0x1626ba7e_u32;
+
+const TRANSACTION_VERSION: felt252 = 1;
+// 2**128 + TRANSACTION_VERSION
+const QUERY_VERSION: felt252 = 340282366920938463463374607431768211457;
+
+#[derive(Serde, Drop)]
+struct Call {
+    to: ContractAddress,
+    selector: felt252,
+    calldata: Array<felt252>
+}
+
+#[abi]
+trait IAccount {
+    #[external]
+    fn __execute__(calls: Array<Call>) -> Array<Array<felt252>>;
+    #[external]
+    fn __validate__(calls: Array<Call>) -> felt252;
+    #[external]
+    fn __validate_declare__(class_hash: felt252) -> felt252;
+    #[external]
+    fn __validate_deploy__(
+        class_hash: felt252, contract_address_salt: felt252, _public_key: felt252
+    ) -> felt252;
+    #[external]
+    fn set_public_key(new_public_key: felt252);
+    #[view]
+    fn get_public_key() -> felt252;
+    #[view]
```

Comment:
> many of these are not part of the standard interface, but non-standard additions

Even if they are not part of the Standard interface, they are part of the contract, and this should represent the interface of the contract, I mean, this Account is not a Standard Account, is the Openzeppelin Account, even if it implements the interface of the Standard Account. This is related to [this comment](https://github.com/OpenZeppelin/cairo-contracts/issues/612#issuecomment-1545653823), I think we should change the terminology for referring to the Standard Account interface.

Reply:
Maybe we could separate the Standard Account interface on another module, and implement it here in (OZ) Account? We would let the full interface for (OZ) Account though.

Reply:
> Maybe we could separate the Standard Account interface on another module, and implement it here in (OZ) Account? We would let the full interface for (OZ) Account though.

This makes the most sense to me. If we go this route, it'd be a good opportunity to use supertraits (when they're available)

Reply:
@martriay WDYT?

Reply:
> I think interfaces ≠ traits. not sure if mixing traits with ABIs/dispatchers is the right design call from the Cairo language, but that may just be my lack of rust knowledge kicking in.
>
> the way I reason about traits is that they allow different types (here, contracts) to share a given logic and interface, guaranteeing a certain behavior over a type (contract), rather than explaining its entire API. for example, I would expect a single contract to implement multiple traits, rather than a single supertrait.
>
>for example, Solidity's IERC20 module doesn't include non-standard functions, and you need to cast in order to call them (e.g. ERC20(address(token)).increaseAllowance)
>
> in short, i would prefer the ABI and dispatcher to be generated from the module and not from the traits, as it's the case with Solidity's contract and interface. am i missing anything? what do you think?

I agree on both your reasoning about traits and the ABI coming from the contract module.

> to decide over this, i would try to think about how do we expect these traits to be used. i expect these traits to be used for contract modules to implement and make sure they're following the Account interface. moreover, i don't want account contracts to be forced to implement non-standard functions because those are present in the standard trait. if it's also the preferred and only way to dispatch functions, we might need multiple traits. then supertraits sound like the way forward.

In think we could separate the concept of the standard interfaces and the ABIs using traits (from what we have now). In the last changes pushed I refactored the Account files to reflect a proposal for handling this, separating the IAccount concept from the ABI one, and splitting the code in multiple files, letting the interface standalone, and the ABI as part of the implementation file. I think it may be easier to understand, what do you think?

Any suggestion for improvements, or a different design would be helpful. cc @andrew-fleming

Reply:
> And I think the naming convention could work in differentiating standard interfaces from non-standard interfaces. Is it enough though? IAccount / AccountABI. Would something like OZ_AccountABI work better? The OZ account is referred to as the "OZ account" anyway. We'd have to be okay with OZ_ERC20ABI and the rest as well.

Very good point. IMO is enough without the OZ prefix, in the same way is being handled for Solidity contracts. [The implementation can use the same name as the standard](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/96b95592c300a1b38953b276f0b6c77017f5c942/contracts/token/ERC20/ERC20.sol#L38), even when it implements different interfaces. In this case, the AccountABI would refer to the ABI of the implementation module (mod Account), not the standard interface.

> Finally, I'm wondering if it makes sense to follow the OZ Solidity repo with having a standalone directory of standard interfaces. The standard IAccount, IERC20, etc sit in interfaces/ (or even standard_interfaces/ for further clarity); whereas, <contract>ABIs reside in the implementing module as it is now. This might be out of scope for this PR, but it's something we can consider

In Solidity, we have the interfaces in the contract directory and then re-exposed in the interfaces directory, with some exceptions (that seem not to be directly implemented). I'm not against this for Cairo, but I think it won't be really helpful until we have some interfaces that are not just re-exposed, because I like the idea of having the interfaces in the contract module as well and having an interfaces directory just for re-exposing seems like unnecessary duplication.

With this said, maybe is better to allow users of our library to use interfaces like this:
```
use openzeppelin::interfaces::IAccount;
```
instead of this?
```
use openzeppelin::account::interface::IAccount;
```
Then the interfaces module would be helpful anyway.

I think I'm still in favor of the second one mostly because users would probably still need to import structs for implementing the interface like Call, and I don't think we should expose Call in openzeppelin::interfaces. So users would still need to do this with the first version:

```
use openzeppelin::interfaces::IAccount;
use openzeppelin::account::interface::Call;
```

and with the second version this:

```
use openzeppelin::account::interface::{IAccount, Call};
```

Reply:
> Very good point. IMO is enough without the OZ prefix, in the same way is being handled for Solidity contracts. [The implementation can use the same name as the standard](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/96b95592c300a1b38953b276f0b6c77017f5c942/contracts/token/ERC20/ERC20.sol#L38), even when it implements different interfaces. In this case, the AccountABI would refer to the ABI of the implementation module (mod Account), not the standard interface.

Mmm that's a fair point. I think you sold me on it :)

> In Solidity, we have the interfaces in the contract directory and then re-exposed in the interfaces directory, with some exceptions (that seem not to be directly implemented). I'm not against this for Cairo, but I think it won't be really helpful until we have some interfaces that are not just re-exposed, because I like the idea of having the interfaces in the contract module as well and having an interfaces directory just for re-exposing seems like unnecessary duplication.

Yeahh, it's definitely better organization to keep relevant code close. The idea was really to further delineate standard and non-standard interfaces and just group the standard interfaces together to ensure they're distinguished, but agreed on all points